packages/buidler-etherscan/src/solc/version.ts

```diff
@@ -0,0 +1,159 @@
+import SemverRange from "semver/classes/range";
+
+import { pluginName } from "../pluginContext";
+
+import {
+  MetadataAbsentError,
+  readSolcVersion,
+  VersionNotFoundError,
+} from "./metadata";
+
+const COMPILERS_LIST_URL =
+  "https://raw.githubusercontent.com/ethereum/solc-bin/gh-pages/bin/list.json";
+
+// Non-exhaustive interface for the official compiler list.
+export interface CompilersList {
+  releases: {
+    [version: string]: string;
+  };
+  latestRelease: string;
+}
+
+export class SolcVersionNumber {
+  constructor(
+    readonly major: number,
+    readonly minor: number,
+    readonly patch: number
+  ) {}
+
+  public async getLongVersion(): Promise<string> {
+    const shortVersion = `${this.major}.${this.minor}.${this.patch}`;
+    const versions = await getVersions();
+    const fullVersion = versions.releases[shortVersion];
+
+    if (fullVersion === undefined || fullVersion === "") {
+      throw new NomicLabsBuidlerPluginError(
+        pluginName,
+        "Given solc version doesn't exist"
+      );
+    }
+
+    return fullVersion.replace(/(soljson-)(.*)(.js)/, "$2");
+  }
+
+  public toString(): string {
+    return `${this.major}.${this.minor}.${this.patch}`;
+  }
+}
+
+export enum InferralType {
+  EXACT,
+  METADATA_PRESENT_VERSION_ABSENT,
+  METADATA_ABSENT,
+}
+
+interface SolcVersionRange {
+  inferralType: InferralType;
+  /**
+   * @returns true if the version is included in the range.
+   */
+  isIncluded(version: SolcVersionNumber): boolean;
+  toString(): string;
+}
+
+export function getVersionNumber(shortVersion: string): SolcVersionNumber {
+  const [major, minor, patch] = shortVersion
+    .split(".", 3)
+    .map((value) => parseInt(value, 10));
+  return new SolcVersionNumber(major, minor, patch);
+}
+
+export async function inferSolcVersion(
+  bytecode: Buffer
+): Promise<SolcVersionRange> {
+  let solcVersionMetadata: SolcVersionNumber;
+  try {
+    solcVersionMetadata = await readSolcVersion(bytecode);
+  } catch (error) {
+    // We want to provide our best inference here.
+    // We can infer that some solidity compiler releases couldn't have produced this bytecode.
+    // Solc v0.4.7 was the first compiler to introduce metadata into the generated bytecode.
+    // See https://solidity.readthedocs.io/en/v0.4.7/miscellaneous.html#contract-metadata
+    // Solc v0.4.26, the last release for the v0.4 series, does not feature the compiler version in its emitted metadata.
+    // See https://solidity.readthedocs.io/en/v0.4.26/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+    // Solc v0.5.9 was the first compiler to introduce its version into the metadata.
+    // See https://solidity.readthedocs.io/en/v0.5.9/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
+    // Solc v0.6.0 features compiler version metadata.
+    // See https://solidity.readthedocs.io/en/v0.6.0/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
```

Comment:
This is a personal preference, but I don't think this complexity is worth it. For me it would be simpler for this function to either return the compiler used, or to not return anything (null or whatever). If the compiler is returned, then we compare with the configured version. If not, we just skip that heuristic and show some warning.

If I'm understanding this correctly, the benefit of all of this is that *if* someone deployed a contract with a version before 0.5.9 they will know whether the compiler used was in the `<0.4.7` range or in the `>=0.4.7 <=0.5.8` range. The value of this seems very low to me compared with the complexity involved.

(A compromise between the two approaches would be to check if the configured compiler satisfies `<0.5.9`, but honestly I think just the warning would be enough.)

@alcuadrado wdyt?